Feat: add k6 version selection to check form

VikaCep opened the pull request:
Adds the channel selector to browser and scripted checks by consuming the new APIs provided by the backend, which are currently mocked.

These new APIs are:
GET `/api/v1/channels/k6`
GET `/api/v1/channel/k6/<channelID>/current`

See https://docs.google.com/document/d/1ZV70lIZ-xOH9EiAAp7OW2LGVeGwgHYJELdcvsSiH7UU/edit?tab=t.1cne0jb892id for a detailed specification of these endpoints and their purpose.

- Displays a new component below the script box that lets users choose the desired k6 version.
<img width="1381" height="1292" alt="image" src="https://github.com/user-attachments/assets/5063ca77-3559-44cb-89bb-ab1be8ac5738" />

- The component consumes the GET `/api/v1/channels/k6` request
<img width="945" height="285" alt="image" src="https://github.com/user-attachments/assets/65b05863-a419-4140-8b23-9dd4737eadfc" />

- When a channel is selected, the GET `/api/v1/channel/k6/<channelID>/current` endpoint is called to obtain the resolved version and be able to display it to the user.
<img width="1071" height="272" alt="image" src="https://github.com/user-attachments/assets/4274d575-4f0e-4255-999f-91b82231e476" />

- When trying to save when a disabled version is selected
> Checks can continue to run past the deprecatedAfter date. At some point after deprecation, we may want to disable the check entirely. Past a channel's disabledDate, checks cannot be updated unless a different channel is picked as part of the update operation. The frontend will show an alert message if a check is using a disabled channel, and potentially also a warning in the checks page if any check is using a disabled channel. 
<img width="1113" height="698" alt="image" src="https://github.com/user-attachments/assets/7f47f030-dccf-4220-9e42-dc8723a0d3eb" />

VikaCep committed:
feat: add feature flag and new types

VikaCep committed:
feat: add component to display k6 versions

VikaCep committed:
feat: mock channel and versions data and integrate with UI components

VikaCep committed:
fix: integrate in check form

VikaCep committed:
fix: pre-select default version

VikaCep committed:
fix: hook with API data

VikaCep committed:
fix: avoid making the channels request when FF is off

VikaCep committed:
fix: handle api call failures

VikaCep committed:
fix: test

ckbedwell commented on src/components/CheckEditor/FormComponents/ChannelDetails.tsx:
```diff
@@ -0,0 +1,82 @@
+
+      {isDeprecated && (
+        <Alert severity="warning" title="Deprecated Channel">
+          This channel is deprecated and will be removed after {new Date(channel.deprecatedAfter).toLocaleDateString()}.
```
After doing the timepoint Explorer I found we have `dateTimeFormat` provided by `@grafana/data` so we don't have to handle this.

VikaCep committed:
fix: add support for disabledAfter prop

VikaCep committed:
fix: use dateTimeFormat

VikaCep committed:
fix: don't show deprecated channels unless previously assigned

VikaCep committed:
fix: don't show disabled channels unless previously assigned

ckbedwell reviewed (changes requested):
Got a few questions and a few things to take a look at 😃 

I don't want to address loading the editor's k6 types in this PR but I am questioning the placement of the channel select. Which channel they select will affect what k6 types we load into the editor, so to me it feels like it would make more sense above the editor rather than below it.

WDYT?

VikaCep committed:
fix: move version selection above script

VikaCep committed:
fix: review comments

VikaCep committed:
fix: add combobox to global ui mocks

VikaCep commented on src/components/CheckEditor/FormComponents/K6ChannelSelect.tsx:
```diff
@@ -0,0 +1,120 @@
+  const { check, isExistingCheck } = useCheckFormMetaContext();
+  const id = 'k6-channel-select';
+
+  const { data: channelsResponse, isLoading: isLoadingChannels, error: channelsError } = useK6Channels(isEnabled);
```
From the backend [design doc](https://docs.google.com/document/d/1ZV70lIZ-xOH9EiAAp7OW2LGVeGwgHYJELdcvsSiH7UU/edit?tab=t.1cne0jb892id):

```
GET /api/v1/channels/k6
{
  "channels": {
    "v0": {
      "name": "v0.x",
      "default": false,
      "deprecatedAfter": "2025-04-01",
      "disabledAfter": "2026-04-01",
      "manifest": "k6>1,k6>=0.53"
    },
    "v1": {
      "name": "v1.x",
      "default": true,
      "deprecatedAfter": "3000-01-01",
      "disabledAfter": "3000-01-01",
      "manifest": "k6>=1"
    },
    "v2": {
      "name": "v2.x",
      "default": false,
      "deprecatedAfter": "3000-01-01",
      "manifest": "k6>=2"
    }
  }
}
```

This API is still in discussion so we could propose to change it to an array if you believe that's more appropriate, so we avoid having to do extra manipulations in the frontend.

VikaCep committed:
fix: improve tests

ckbedwell reviewed (dismissed):
Love it 👏

VikaCep committed:
fix: add zod validation for selected disabled channels

VikaCep committed:
test: add payload tests for the new channel prop

VikaCep commented:
@ckbedwell I added some changes related to validation based on some changes on the backend design doc:
Now, besides `deprecatedAfter`, channels provide the `disabledAfter` prop. If a check is assigned a channel that is disabled, the dropdown will include it in the list, it but we don't allow saving the form until that's changed. From the doc:

> Checks can continue to run past the deprecatedAfter date. At some point after deprecation, we may want to disable the check entirely. Past a channel's disabledDate, checks cannot be updated unless a different channel is picked as part of the update operation. The frontend will show an alert message if a check is using a disabled channel

If this happens, we now show this validation error message:
<img width="2226" height="1396" alt="image" src="https://github.com/user-attachments/assets/54187b30-91de-470c-909b-613c8e9aaeb5" />

I also added some payload tests that were missing.

As a sidenote, I started thinking about loading the correct k6 types for the editor but I think it will be better to address in a different PR. I have an issue for it here: https://github.com/grafana/synthetic-monitoring-app/issues/1252

ckbedwell commented on src/components/CheckEditor/FormComponents/ChannelDetails.tsx:
```diff
@@ -0,0 +1,73 @@
+      <Text variant="bodySmall" color="secondary">
+        k6 version constraint:{' '}
+        <code
+          style={{
+            backgroundColor: theme.colors.background.secondary,
+            color: theme.colors.text.primary,
+            padding: '2px 6px',
+            borderRadius: theme.shape.radius.default,
+            fontSize: theme.typography.bodySmall.fontSize,
+            fontFamily: theme.typography.fontFamilyMonospace,
+          }}
```
Nit: this is minor but can we use a `className` here instead?

ckbedwell reviewed (changes requested):
With these new changes and their validation considerations it has become apparent that the k6 channels affect the whole scripted and browser form experience so I think we should invest some effort into making them more integrated into the schema from a higher level rather than overloading the `<K6ChannelSelect />` component with too many considerations.

VikaCep committed:
fix: move channel under settings and change to array format

VikaCep committed:
refactor: lift k6 channel validation from component to schema

- Move validation logic from K6ChannelSelect useEffect to Zod schemas
- Create schema factory functions for browser/scripted checks
- Lift useK6Channels hook to page level components
- Remove channelDisabled form field in favor of schema validation
- Add graceful error handling for k6 channels API failures

VikaCep committed:
fix: fixtures

VikaCep committed:
fix: channel error path

VikaCep committed:
fix: deprecation message

VikaCep committed:
fix: minor

VikaCep committed:
refactor: move k6 channels filtering and default logic to data/transformation layers

VikaCep committed:
fix: tests

VikaCep committed:
fix: test

VikaCep committed:
fix: tests

VikaCep committed:
fix: tests

VikaCep committed:
fix: remove unused checkType param from toFormValues

VikaCep committed:
fix: remove channel from baseCheckSchema

VikaCep committed:
fix: add channel to scripted/browser fields object

VikaCep committed:
fix: add comments and new tests on showing default k6 version message

VikaCep committed:
fix: db import for dev:msw to work

VikaCep committed:
fix: use filtered channels from component pages to prevent repetition

VikaCep committed:
fix: allow to set channel when /current fails

VikaCep committed:
fix: only include channel in payload when it has a value

VikaCep committed:
fix: remove channel from check's base model

VikaCep committed:
fix: port to new checkform

VikaCep committed:
fix: tests

VikaCep committed:
fix: remove deprecatedAfter prop and usage

As discussed with the backend, this is not going to be supported

VikaCep committed:
fix: remove current version references

As discussed with the backend, this won't be supported

VikaCep committed:
fix: lint

VikaCep committed:
fix: add missing transformations for the new editor and migrate tests

VikaCep committed:
fix: remove k6channels from context as it's no longer needed

VikaCep committed:
fix: tests
